### PR TITLE
Fix Room Spek Custom Data Tests

### PR DESCRIPTION
**Pull request by daniellevass**
What
Fixed the room custom data tests as they were pretty flaky and refactored into their own file to reduce how big room spek is getting!

Why
Flaky tests are not great for us to know if things are actually broken :) This should fix that!

### CODE REVIEW

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            val newCustomData = mapOf(
+                    "added" to "some",
+                    "custom" to "data"
+            )
+
+            var alicesUpdatedGeneralRoom by FutureValue<Room>()
+            chatFor(ALICE).connectFor { event ->
+                when (event) {
+                    is ChatEvent.RoomUpdated -> {
+                        alicesUpdatedGeneralRoom = event.room
```
Comment (daniellevass) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: make > `alicesRoomUpdatedEvent`

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt [hunk 2]
```diff
@@ -83,32 +85,40 @@ object RoomSpek : Spek({
             assertThat(updatedRoom.name).isEqualTo(NOT_GENERAL)
         }
 
-        it("notifies '$ALICE' when room '$GENERAL' receives a new message") {
-            setUpInstanceWith(createDefaultRole(), newUsers(ALICE, PUSHERINO), newRoom(GENERAL, ALICE, PUSHERINO))
+        it("updates the unread counter when '$GENERAL' receives a new message") {
```
Comment (daniellevass) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt]: make these tests the same again

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+import com.pusher.chatkit.Users.PUSHERINO
+import com.pusher.chatkit.Users.SUPER_USER
+import com.pusher.chatkit.test.InstanceActions.createDefaultRole
+import com.pusher.chatkit.test.InstanceActions.newRoom
+import com.pusher.chatkit.test.InstanceActions.newUsers
+import com.pusher.chatkit.test.InstanceSupervisor.setUpInstanceWith
+import com.pusher.chatkit.test.InstanceSupervisor.tearDownInstance
+import com.pusher.chatkit.util.FutureValue
+import junit.framework.Assert.assertNotNull
+import org.jetbrains.spek.api.Spek
```
Comment (marekoid) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: Where there finally any issues in using this move/partial-rewrite as an opportunity to bring Spek 2 here? Just asking as I remember you were enthusiastic about doing it :)

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+import org.jetbrains.spek.api.dsl.describe
+import org.jetbrains.spek.api.dsl.it
+
+object RoomCustomDataSpek : Spek({
```
Comment (marekoid) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: For new tests I was using clear naming. How does `RoomCustomDataE2eTest` sound? I found it easier to navigate when tests are clearly named by their type and actually called test instead of naming them after the framework used (plus Spec vs Spek — for actual names in the docs they use `c`).

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    beforeEachTest(::tearDownInstance)
```
Comment (marekoid) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: It's more clear to have tearDown actually after test. It also leaves the instance clean at the end. Having it after `closeChatManagers` would be good.

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt [hunk 6]
```diff
@@ -83,32 +85,26 @@ object RoomSpek : Spek({
             assertThat(updatedRoom.name).isEqualTo(NOT_GENERAL)
         }
 
-        it("notifies '$ALICE' when room '$GENERAL' receives a new message") {
-            setUpInstanceWith(createDefaultRole(), newUsers(ALICE, PUSHERINO), newRoom(GENERAL, ALICE, PUSHERINO))
+        it("updates the unread count and last message at when '$GENERAL' receives a new message") {
```
Comment (marekoid) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt]: this test is still changed, whilst the [intention ](https://github.com/pusher/chatkit-android/pull/148#discussion_r370208336) was to revert the changes

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt (hunk 4 above)
Comment (daniellevass) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: Hmm. I'm not sure how I feel about having `E2e`. The other tests are currently `*Spek` so I don't want to add too much confusion there. Let's talk about this for v2 on what kind of naming schemes we want for to use and keep so we can be as consistent?

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt (hunk 5 above)
Comment (daniellevass) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: We already have `afterEachTest(::closeChatManagers)` which is L22 - is that what you meant? I kept whatever we had existing in the `RoomSpek` tests.

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt (hunk 6 above)
Comment (daniellevass) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt]: I think I may have misunderstood - i thought we wanted to combine the two tests - are you asking my to revert the change that test and keeping what we had previously? If so, we can just abandon this PR?

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                    is ChatEvent.RoomUpdated -> {
+                        alicesRoomUpdatedEvent = event
+                    }
+                }
+            }
+
+            superUser.updateRoom(
+                    room = superUser.generalRoom,
+                    customData = newCustomData
+
```
Comment (marekoid) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: the whole call  looks a bit disjoined now with this added empty line

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            ).assumeSuccess()
+
+            assertThat(alicesRoomUpdatedEvent.room.customData).isEqualTo(newCustomData)
+            assertThat(superUserRoomUpdatedEvent.room.customData).isEqualTo(newCustomData)
+            assertNotNull(superUser.rooms[0].customData)
+            assertThat(superUser.rooms[0].customData).isEqualTo(newCustomData)
```
Comment (marekoid) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: for completeness it would be good to have the same check for alice, grouping them event assertion + state assertion for each user and separating with an empty line

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+            superUser.updateRoom(
+                    room = superUser.generalRoom,
+                    customData = newCustomData
+
+            ).assumeSuccess()
+
+            assertThat(alicesRoomUpdatedEvent.room.customData).isEqualTo(newCustomData)
+            assertThat(superUserRoomUpdatedEvent.room.customData).isEqualTo(newCustomData)
+            assertNotNull(superUser.rooms[0].customData)
```
Comment (marekoid) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: adding that line is not needed, the existing assertion below already achieves the same

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt (hunk 6 above)
Comment (marekoid) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt]: The test used to combine the two checks (lastMessageAt and unreadCount) and that's needed as we receive 2 RoomUpdated events for each property and don't have guarantee on the order of them.

Now the test is just checking lastMessageAt event and not as it used to i.e. not from the perspective of the message recipient (which is more important than the sender's perspective). Also it uses junit assertions as opposed to Truth.

Let's revert the changes in this test.

The extraction and fixing of flakiness in custom data test is definitely a nice thing, not related to the ME effort but as it got discovered as part of it, I would be keen on getting it merged.

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt (hunk 5 above)
Comment (marekoid) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: That's what I mean:
```kotlin
    afterEachTest(::closeChatManagers)
    afterEachTest(::tearDownInstance)
```

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt (hunk 4 above)
Comment (marekoid) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: The existing tests were named as "Spec", "Spek" or "Test". I added quite a few named as "Test".

I think that `E2e` is the most descriptive name. I was also thinking about "System" but thought that it's better to keep that name for OS integration tests (UI Automator) if we create any in the future. "End to end" also conveys the information more clearly than "system".

Fair enough if you are not convinced and don't feel like changing it now ;) I just wanted to use this opportunity for that new(ish) test to be named consistently with `FunctionalTest`s that has been recently created.

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt (hunk 8 above)
Comment (marekoid) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: not sure if I got understood as I meant grouping by user but groping by type is fine too so whatever you prefer :)
```kotlin
            assertThat(alicesRoomUpdatedEvent.room.customData).isEqualTo(newCustomData)
            assertThat(alice.rooms[0].customData).isEqualTo(newCustomData)
            
            assertThat(superUserRoomUpdatedEvent.room.customData).isEqualTo(newCustomData)
            assertThat(superUser.rooms[0].customData).isEqualTo(newCustomData)
```

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt (hunk 5 above)
Comment (daniellevass) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: Ah, doing both things after each test :) That makes sense. Have refactored!

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt (hunk 6 above)
Comment (daniellevass) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt]: Okay, i'll update the title of this PR to improve clarity around what's actually happening now.

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt (hunk 4 above)
Comment (daniellevass) [reply chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomCustomDataSpek.kt]: Have compromised and renamed it to `Test` :) We can talk about what kind of naming convention we want for v2 and start renaming things 😄

**Comment by daniellevass** [comment]
Have resolved both those pieces of feedback and renamed the PR title to reflect more accurately what this change actually is.

File: chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt [hunk 10]
```diff
@@ -84,9 +86,15 @@ object RoomSpek : Spek({
...
-            setUpInstanceWith(createDefaultRole(), newUsers(ALICE, PUSHERINO), newRoom(GENERAL, ALICE, PUSHERINO))
+            setUpInstanceWith(
+                    createDefaultRole(),
+                    newUsers(ALICE, PUSHERINO),
+                    newRoom(GENERAL, ALICE, PUSHERINO))
 
             val alice = chatFor(ALICE).connect().assumeSuccess()
+            val pusherino = chatFor(PUSHERINO).connect().assumeSuccess()
+
+
```
Comment (marekoid) [inline chatkit-core/src/integrationTest/kotlin/com/pusher/chatkit/RoomSpek.kt]: redundant empty line?
